### PR TITLE
Field of View and Blindness improvements [bounty + upstream push]

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request
I have received exemption from the freeze by Oranges.
This PR was originally done for Skyrat-tg, as a bounty. Oranges took interest and Skyrat people approved taking it upstream so I did. (original PR https://github.com/Skyrat-SS13/Skyrat-tg/pull/9912 )
Features:
This PR adds "field of view" mechanics, which by default don't do anything.
Once you pick up and wear a gas mask you'll notice a cone behind you.
That cone will hide all mobs, projectiles and projectile effects.
HOWEVER... hidden things in the cone (or for people who are blind or nearsighted) will play a visual effect.
FOV cone:
dreamseeker_2021-12-08_21-12-14
Footsteps:
dreamseeker_2021-12-08_20-43-43
Gunfire:
dreamseeker_2021-12-08_22-43-00
Attacks:
dreamseeker_2021-12-08_22-45-38

The gas mask tint is probably gonna get removed in favor of this (expect this in another PR?)
obraz

The fov cone shadow density can be adjusted

Technicalities:
~~Originally a component, but by Watermelon's advice I have made this a datumized behaviour instead. (semi-component) This way it is faster, less memory use etc.~~ Once again a component
People have native fov, which can be enabled by a config, by default it's off. (or by an admin verb)
People have fov traits, which add the fov impairings to you, gas masks do that now. All clothes have a support for possibly applying them.
Moves all things that should be hidden onto a new plane which is just above GAME_PLANE, that was we can efficiently filter them out of clients views.
Being dead or extending your eye view (xenobio console) will disable the FOV mask.

This PR was brought to you by Skyrat paying me aswell as rotational mathematics
mspaint_2021-12-08_03-15-26

## Why It's Good For The Game
Oranges approves so it's good.

## Changelog

:cl:
add: Gas Masks now limit your field of view
add: Added field of view mechanics. It's darkness can be tweaked by a pref
add: Blind, nearsighted and fov-limited people will now see sounds as visual effects.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/living_defines.dm [hunk 1]
```diff
@@ -171,3 +172,12 @@
 	var/body_position_pixel_x_offset = 0
 	///The x amount a mob's sprite should be offset due to the current position they're in
 	var/body_position_pixel_y_offset = 0
+
+	/// FOV view that is applied from either nativeness or traits
+	var/fov_view
+	/// Native FOV that will be applied if a config is enabled
+	var/native_fov = FOV_90_DEGREES
```
Comment (optimumtact) [inline code/modules/mob/living/living_defines.dm:179]: if you have a datum, why are you applying these settings on the mob level

File: code/modules/mob/living/living_fov.dm [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	blocker_mask.dir = new_dir
+	visual_shadow.dir = new_dir
+
+/datum/field_of_view/proc/set_fov_angle(new_angle)
+	fov_angle = new_angle
+	blocker_mask.icon_state = "[fov_angle]"
+	visual_shadow.icon_state = "[fov_angle]_v"
+
+/// Is `observed_atom` in a mob's field of view? This takes blindness, nearsightness and FOV into consideration
+/mob/living/proc/in_fov(atom/observed_atom, ignore_self = FALSE)
```
Comment (optimumtact) [inline code/modules/mob/living/living_fov.dm:129]: why is this a proc on the mob and not on the datum (taking a source atom).

File: code/modules/mob/living/living_defines.dm (hunk 1 above)
Comment (Azarak) [reply code/modules/mob/living/living_defines.dm:179]: The datum is the handler for the client, and is only instantiated when a client enters that mob. 
You will want to manage those variables whether that handler exists or not.

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { multiline } from "common/string";
+import { CheckboxInput, FeatureToggle } from "../base";
+
+export const play_fov_effects: FeatureToggle = {
```
Comment (optimumtact) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx:4]: why is this a preference?

File: code/modules/mob/living/living_fov.dm (hunk 2 above)
Comment (Azarak) [reply code/modules/mob/living/living_fov.dm:129]: In previous reply I clarified that the datum is the client handler. You may want to check `in_fov()` whether that mob has a client with a fov handler or not. Currently this being on mob/living isn't nessecary as we only use it to play the effects which already require a client, but for any expansions upon the system I imagine this will be useful.

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx (hunk 3 above)
Comment (Azarak) [reply tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx:4]: This is the preference for the visual effects inside FOV being played (footsteps, attacks). Should it be not? I can remove it

File: code/modules/mob/living/living.dm [hunk 4]
```diff
@@ -10,6 +10,7 @@
 	faction += "[REF(src)]"
 	GLOB.mob_living_list += src
 	SSpoints_of_interest.make_point_of_interest(src)
+	update_fov()
```
Comment (optimumtact) [inline code/modules/mob/living/living.dm:13]: why does this not dispatch to the fov handler

Comment (Azarak) [reply code/modules/mob/living/living.dm:13]: Instantiated mob might not get the handler because it may not have a client. Determining a thing's FOV despite it being clientless could be important for any expansions upon the system, but currently it isn't important and could be updated when the handler is about to be created, I can change it if need be.

File: code/modules/mob/living/living_fov.dm [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+	// If turf doesn't exist, then we wouldn't get a fov check called by `play_fov_effect` or presumably other new stuff that might check this.
+	//  ^ If that case has changed and you need that check, add it.
+	var/rel_x = observed_atom.x - my_turf.x
+	var/rel_y = observed_atom.y - my_turf.y
+	if(fov_view)
+		if(rel_x >= -1 && rel_x <= 1 && rel_y >= -1 && rel_y <= 1) //Cheap way to check inside that 3x3 box around you
+			return TRUE //Also checks if both are 0 to stop division by zero
+	
+		// Converts the relative position into a 0-360 rotation
+		var/vector_len = sqrt(abs(rel_x) ** 2 + abs(rel_y) ** 2)
```
Comment (optimumtact) [inline code/modules/mob/living/living_fov.dm:145]: if this is a 0-360 rotation why is the var name vector_len

Comment (Azarak) [reply code/modules/mob/living/living_fov.dm:145]: Oh my bad. I changed my method and forgot to update the comment. Will fix

File: code/modules/mob/living/living_fov.dm [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	
+		/// Calculate vision angle and compare
+		var/vision_angle = (360 - fov_view) / 2
+		if(angle < vision_angle)
+			. = TRUE
+	else
+		. = TRUE
+
+	// Handling nearsightnedness
+	if(. && HAS_TRAIT(src, TRAIT_NEARSIGHT))
```
Comment (SuperNovaa41) [inline code/modules/mob/living/living_fov.dm:175]: this should probably be an early return

Comment (Azarak) [reply code/modules/mob/living/living_fov.dm:175]: I dont think there should be? There are checks we have to run to determine whether someone is really nearsighted, and if that person is far away to receive those effects

**Comment by Mosslick** [comment]
Why would you bundle something awesome as blindness improvements with something absolute garbage as even MORE gas mask nerfs?
Nobody will ever use gas masks now they are literally useless, why the hell are these two bundled

**Comment by Azarak** [comment]
> Why would you bundle something awesome as blindness improvements with something absolute garbage as even MORE gas mask nerfs? Nobody will ever use gas masks now they are literally useless, why the hell are these two bundled

Well the blindness improvements were aimed for FOV (as this was for a downstream who wants native FOV).
...but fret not... I heard a rumor from a trusted source the gasmask tint will be going away, so it'll be just the FOV

File: code/modules/mob/living/living_fov.dm (hunk 6 above)
Comment (SuperNovaa41) [reply code/modules/mob/living/living_fov.dm:175]:
```suggestion
	if(!. || !HAS_TRAIT(src, TRAIT_NEARSIGHT))
	    return
	... the rest of the code but on one less indentation level
```

**Comment by D4C-420** [comment]
This is going to suck for miners, good luck fighting megafauna  when they are hidden and off your screen half the time leaving you with no good way to tell what attack they are going to do or even where they are.

**Comment by Azarak** [comment]
> This is going to suck for miners, good luck fighting megafauna when they are hidden and off your screen half the time leaving you with no good way to tell what attack they are going to do or even where they are.

You can use a breath mask instead of a gas mask. And I suppose I could make it so simple mobs dont get hidden, or that megafaunas dont get hidden, and I will should any maintainer would want it that way

**Comment by Fikou** [comment]
> You can use a breath mask instead of a gas mask. And I suppose I could make it so simple mobs dont get hidden, or that megafaunas dont get hidden, and I will should any maintainer would want it that way

you dont get a breath mask as a miner, you only get mining gas masks in both your survival box and in your suit storage units, you are intended to wear them the entire round

File: code/modules/mob/living/living_fov.dm (hunk 6 above)
Comment (Azarak) [reply code/modules/mob/living/living_fov.dm:175]: I suppose it can but wouldn't that be less readable?

Comment (Mothblocks) [reply code/modules/mob/living/living_fov.dm:175]: It shouldn't be an early-return, early-returns are for conditionals that are required for the process to work, generally. Having a lot of different conditionals not early-returned is fine, it means it's easy to add more.

**Comment by carshalash** [comment]
Wait, why freeze exempt?

File: code/modules/mob/living/living_fov.dm [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+	if(!parent_client) //Love client volatility!!
+		qdel(src)
+		return
+
+	fov_angle = fov_type
+	blocker_mask = new
+	visual_shadow = new
+	blocker_mask.dir = owner.dir
+	visual_shadow.dir = owner.dir
+	visual_shadow.alpha = parent_client.prefs.read_preference(/datum/preference/numeric/fov_darkness)
```
Comment (Mothblocks) [inline code/modules/mob/living/living_fov.dm:35]: Will need to check client here too

Comment (Mothblocks) [reply code/modules/mob/living/living_fov.dm:35]:
```suggestion
	visual_shadow.alpha = parent_client?.prefs.read_preference(/datum/preference/numeric/fov_darkness)
```

File: code/modules/mob/living/living_fov.dm [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+	if(visual_shadow) // In a case of early deletion due to volatile client
+		QDEL_NULL(visual_shadow)
+	return ..()
+
+/// Updates the state of the owner living and updates masking
+/datum/field_of_view/proc/update_living()
+	if(owner.stat == DEAD)
+		user_living = FALSE
+	else
+		user_living = TRUE
```
Comment (Mothblocks) [inline code/modules/mob/living/living_fov.dm:56]:
```suggestion
	user_living = owner.stat != DEAD
```

File: code/modules/mob/living/living_fov.dm [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+	update_masking()
+
+/// Updates the state of the owner eye and updates masking
+/datum/field_of_view/proc/update_eye()
+	var/client/parent_client = owner.client
+	if(parent_client.eye == owner)
+		user_extends_eye = FALSE
+	else
+		user_extends_eye = TRUE
```
Comment (Mothblocks) [inline code/modules/mob/living/living_fov.dm:65]:
```suggestion
	user_extends_eye = owner.client.eye != owner
```

File: code/modules/mob/living/living_fov.dm [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+	update_masking()
+
+/// Properly applies or removes the mask, on conditions of living and not extending an eye
+/datum/field_of_view/proc/update_masking()
+	var/should_apply_mask = FALSE
+	if(user_living && !user_extends_eye)
+		should_apply_mask = TRUE
```
Comment (Mothblocks) [inline code/modules/mob/living/living_fov.dm:72]:
```suggestion
	var/should_apply_mask = user_living && !user_extends_eye
```

File: code/modules/mob/living/living_fov.dm [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+		add_mask()
+	else
+		remove_mask()
+
+/// Updates the size of the FOV masks by comparing them to client view size.
+/datum/field_of_view/proc/update_fov_size()
+	var/client/parent_client = owner.client
+	if(!parent_client) //Love client volatility!!
+		return
+	var/list/view_size = getviewsize(parent_client.view)
```
Comment (Mothblocks) [inline code/modules/mob/living/living_fov.dm:85]: Check here too

File: code/modules/mob/living/logout.dm [hunk 12]
```diff
@@ -4,3 +4,6 @@
 	if(!key && mind) //key and mind have become separated.
 		mind.active = FALSE //This is to stop say, a mind.transfer_to call on a corpse causing a ghost to re-enter its body.
 	med_hud_set_status()
+
+	if(fov_handler)
+		qdel(fov_handler)
```
Comment (Mothblocks) [inline code/modules/mob/living/logout.dm:9]:
```suggestion
	QDEL_NULL(fov_handler)
```

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx (hunk 3 above)
Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx:4]: I don't see any reason for it to be a preference no

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/fov_darkness.tsx [hunk 13]
```diff
@@ -0,0 +1,7 @@
+import { Feature, FeatureNumberInput } from "../base";
+
+export const fov_darkness: Feature<number> = {
+  name: "FoV Darkness",
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/fov_darkness.tsx:4]: Please write out the full name and also a description of what this is, I didn't even know what FoV was referring to when it was being discussed. Like, yes, I know it's field of view, but not how that related to the game.

File: code/modules/client/preferences/fov_darkness.dm [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+	savefile_key = "fov_darkness"
+	savefile_identifier = PREFERENCE_PLAYER
+
+	minimum = 0
+	maximum = 255
+
+/datum/preference/numeric/fov_darkness/create_default_value()
+	return 255
+
+/datum/preference/numeric/fov_darkness/apply_to_client(client/client, value)
```
Comment (Mothblocks) [inline code/modules/client/preferences/fov_darkness.dm:12]: This should likely be `apply_to_client_updated`?

File: code/modules/mob/living/death.dm [hunk 15]
```diff
@@ -91,6 +91,9 @@
 	med_hud_set_status()
 	stop_pulling()
 
+	if(fov_handler)
+		fov_handler.update_living()
```
Comment (Mothblocks) [inline code/modules/mob/living/death.dm:95]:
```suggestion
	fov_handler?.update_living()
```

File: code/modules/mob/living/living_fov.dm (hunk 7 above)
Comment (Azarak) [reply code/modules/mob/living/living_fov.dm:35]: Think so? The client was just checked above.

File: code/__DEFINES/fov.dm [hunk 16]
```diff
@@ -0,0 +1,11 @@
+/// Field of vision defines.
+#define FOV_90_DEGREES 90
+#define FOV_180_DEGREES 180
+#define FOV_270_DEGREES 270
```
Comment (Mothblocks) [inline code/__DEFINES/fov.dm:4]: Why are these defines? When will they change? Their value is already embedded in their name

File: code/datums/elements/footstep.dm [hunk 17]
```diff
@@ -134,6 +134,7 @@
 	if(!source_loc)
 		return
 
+	play_fov_effect(source, 5, "footstep", direction, TRUE)
```
Comment (Mothblocks) [inline code/datums/elements/footstep.dm:137]: Named param for the `TRUE`

File: code/modules/admin/admin_verbs.dm [hunk 18]
```diff
@@ -105,6 +105,7 @@ GLOBAL_LIST_INIT(admin_verbs_fun, list(
 	/client/proc/smite,
 	/client/proc/admin_away,
 	/datum/admins/proc/station_traits_panel,
+	/client/proc/cmd_admin_toggle_fov
```
Comment (Mothblocks) [inline code/modules/admin/admin_verbs.dm:108]:
```suggestion
	/client/proc/cmd_admin_toggle_fov,
```

File: code/modules/client/client_procs.dm [hunk 19]
```diff
@@ -1033,6 +1033,8 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 	if (isliving(mob))
 		var/mob/living/M = mob
 		M.update_damage_hud()
+		if(M.fov_handler)
+			M.fov_handler.update_fov_size()
```
Comment (Mothblocks) [inline code/modules/client/client_procs.dm:1037]:
```suggestion
		M.fov_handler?.update_fov_size()
```

File: code/modules/client/preferences/ambient_occlusion.dm [hunk 20]
```diff
@@ -10,3 +10,9 @@
 		return
 
 	plane_master.backdrop(client.mob)
+
+	var/atom/movable/screen/plane_master/game_world/fov_hidden/plane_master_fov = locate() in client?.screen
+	if (!plane_master_fov)
+		return
+
+	plane_master_fov.backdrop(client.mob)
```
Comment (Mothblocks) [inline code/modules/client/preferences/ambient_occlusion.dm:18]: What does this have to do with ambient occlusion?

File: code/modules/clothing/clothing.dm [hunk 21]
```diff
@@ -38,6 +38,9 @@
 	/// Trait modification, lazylist of traits to add/take away, on equipment/drop in the correct slot
 	var/list/clothing_traits
 
+	/// Whether the clothing applies a fov trait, which limits a users field of view.
+	var/fov_trait
```
Comment (Mothblocks) [inline code/modules/clothing/clothing.dm:42]: This should be a component or an element, not a var

**Review by Mothblocks — CHANGES_REQUESTED** [review]
The datum makes me uneasy, it feels like a lot of it should just be a component.

`human.AddComponent(/datum/component/field_of_view, 270)` is what seems ideal in my head?

File: code/modules/mob/living/living_fov.dm (hunk 7 above)
Comment (Mothblocks) [reply code/modules/mob/living/living_fov.dm:34]: It's precisely your comment about clients--from both me and oranges' experience, `/client` can delete whenever, wherever.

**Comment by Mothblocks** [comment]
@carshalash Cuz oranges asked for it or something

File: code/__DEFINES/fov.dm (hunk 16 above)
Comment (Azarak) [reply code/__DEFINES/fov.dm:4]: These exist so people don't use a value that doesn't have a corresponding mask in the .dmi file

File: code/modules/mob/living/logout.dm (hunk 12 above)
Comment (Azarak) [reply code/modules/mob/living/logout.dm:9]: fov_handler nulls the reference to itself on Destroy()

File: code/modules/client/preferences/ambient_occlusion.dm (hunk 20 above)
Comment (Azarak) [reply code/modules/client/preferences/ambient_occlusion.dm:18]: It creates a new plane which requires the ambient occlusion to mimick looking just like the game plane. If there is a way to bundle up the planes and apply that backdrop together to them then I'll do it, but I do not know how or whether it's possible. I'm not familiar with the new rendering capabilities after the refactor.

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx (hunk 3 above)
Comment (optimumtact) [reply tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx:4]: yeah why on earthw ould you ever disable it, no need.

**Comment by SkeletalElite** [comment]
This is way cooler than a silly darkening tint

File: code/modules/mob/living/logout.dm (hunk 12 above)
Comment (LemonInTheDark) [reply code/modules/mob/living/logout.dm:9]: do you trust it?

Comment (Azarak) [reply code/modules/mob/living/logout.dm:9]: I guess it's 1 line less

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx (hunk 3 above)
Comment (Azarak) [reply tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/play_fov_effects.tsx:4]: I shall remove it then

**Comment by bluetshirtguy** [comment]
FOV is a gimmick like pixel movement.

**Comment by LemonInTheDark** [comment]
> The datum makes me uneasy, it feels like a lot of it should just be a component.
> 
> `human.AddComponent(/datum/component/field_of_view, 270)` is what seems ideal in my head?

I agree with this, at least in principle. I think the value should be tracked outside of a component as source -> value, and the component should just handle the visuals, but eh

**Comment by Mothblocks** [comment]
Tracking the value outside of a component is okay, as long as the total code surface area is kept minimal.

**Comment by cacogen** [comment]
I remember errorage tried something like this a long time ago but his approach involved moving the camera ahead of where the mob was facing in order to hide everything behind it offscreen so it didn't work very well. Could be interesting, would have to try it out first to know though. I like the effects to make blindness less miserable at least

File: code/__DEFINES/fov.dm (hunk 16 above)
Comment (stylemistake) [reply code/__DEFINES/fov.dm:4]: Where is a define for 0 degrees, then

**Comment by Stalkeros2** [comment]
Reposting my commentary from the OG PR.

Combat mode is already "oversaturated" with functions and is crucial in X or Y department's/person's gameplay (I myself just randomly flip it on or off when I am very bored so you know that might result in "Schrödinger's people" existing. Besides, mining usually involves changing your intents constantly. And I'm a miner. Flickering miner.), so setting it to the combat mode isn't the best idea tbh.
I'd recommend making a separate stealth mode, just like me and some other folks in Discord had suggested.

**Comment by Time-Green** [comment]
The fucking madman actually did it

**Comment by LemonInTheDark** [comment]
> 

Read the pr desc, we're not taking the combat mode thing because it's not at all suited to tg

**Comment by necromanceranne** [comment]
I kinda expected this PR and I am actually way more on board with this. But  eing coupled with blindness changes is great as well! Loving the possibilities!

**Comment by Stalkeros2** [comment]
> > Reposting my commentary from the OG PR.
> > Combat mode is already "oversaturated" with functions and is crucial in X or Y department's/person's gameplay (I myself just randomly flip it on or off when I am very bored so you know that might result in "Schrödinger's people" existing. Besides, mining usually involves changing your intents constantly. And I'm a miner. Flickering miner.), so setting it to the combat mode isn't the best idea tbh. I'd recommend making a separate stealth mode, just like me and some other folks in Discord had suggested.
> 
> Read the pr desc, we're not taking the combat mode thing because it's not at all suited to tg

Oh good.

File: code/modules/clothing/clothing.dm (hunk 21 above)
Comment (Ghommie) [reply code/modules/clothing/clothing.dm:42]: This. It's doable.

Comment (Azarak) [reply code/modules/clothing/clothing.dm:42]: Yeah I'll get on that in a bit.

**Comment by Azarak** [comment]
I'll move the clothing fov trait into an element in a bit.
What's the consensus on whether `fov_handler` should be a component or a datum like it's now? @optimumtact @Mothblocks 
It's doesn't get instantiated like most components do, because it disappears and reappears with client presence.

**Review by Ghommie — COMMENTED** [review]
Shouldn't vehicles also have the `GAME_PLANE_FOV_HIDDEN` plane? In the current state they could pretty much give out the current location of mobs and players who are supposed to be hidden by the FOV.

**Comment by Mothblocks** [comment]
@Azarak I don't care if it's a component or a datum so long as its surface area is low. Currently `fov_handler?.do_thing()` is called in several base procs, which is precisely the problem signals are meant to solve.

**Comment by Dawson1917** [comment]
FOV implementation is very cool, especially with the effort to give cues for various actions happening behind you, but I think it would just be too annoying for SS13 with our current control scheme. It's too clunky to move your character around to where your eyes are already looking. I'd take annoyingly tinted screens over infuriating controls any day.

Unless... The FOV cone was made to move with the mouse, like in other games..? And the player sprite can move to face it the same way it does when you examine something at a distance..? Just spitballing.

**Comment by LemonInTheDark** [comment]
If we add more things to mouseentered I will explode into a pile of ash

**Comment by Mothblocks** [comment]
We have no client side

**Comment by Striders13** [comment]
> > This is going to suck for miners, good luck fighting megafauna when they are hidden and off your screen half the time leaving you with no good way to tell what attack they are going to do or even where they are.
> 
> You can use a breath mask instead of a gas mask. And I suppose I could make it so simple mobs dont get hidden, or that megafaunas dont get hidden, and I will should any maintainer would want it that way

Yeah you can't really use a breath mask, if you get set of fire (which happens quite a lot when mining) breath mask will burn and you'll just get insta knock out by co2 and/or toxin damage from plasma. Doesn't sound fun.

Also I'm interested if blind people can see simplemob footsteps?

**Comment by Stalkeros2** [comment]
Use GPSes to detect megafauna and memorise their attack patterns.
So.
Skill issue.
:sunglasses:

**Comment by drIcepop** [comment]
If this change does go through(I'd rather we just use old gas masks, as there was literally nothing wrong with them), I don't think this should apply to miner and sec masks.

**Comment by SoreYew** [comment]
Concerns about miners(and to a lesser extent atmosians, and maybe security) losing vision is more of a balance,than the core of the feature, isnt it? A generic emergency use gas mask and one made for explicitly hostile environments could surely have visor differences as they do irl, for example.

**Comment by Azarak** [comment]
> Shouldn't vehicles also have the `GAME_PLANE_FOV_HIDDEN` plane? In the current state they could pretty much give out the current location of mobs and players who are supposed to be hidden by the FOV.

Vehicles are now hidden, thanks for the suggestion.

**Comment by Azarak** [comment]
> > > This is going to suck for miners, good luck fighting megafauna when they are hidden and off your screen half the time leaving you with no good way to tell what attack they are going to do or even where they are.
> > 
> > 
> > You can use a breath mask instead of a gas mask. And I suppose I could make it so simple mobs dont get hidden, or that megafaunas dont get hidden, and I will should any maintainer would want it that way
> 
> Yeah you can't really use a breath mask, if you get set of fire (which happens quite a lot when mining) breath mask will burn and you'll just get insta knock out by co2 and/or toxin damage from plasma. Doesn't sound fun.
> 
> Also I'm interested if blind people can see simplemob footsteps?

Simplemob footsteps do not make the effects for blind people. I can change that should a maintainer want. I didn't really see much impact from that, but possibly you could have a bunch more idle CPU usage from a lot of simplemobs on away levels stumbling around.

File: code/modules/clothing/clothing.dm (hunk 21 above)
Comment (Azarak) [reply code/modules/clothing/clothing.dm]: And it is now!

**Comment by Azarak** [comment]
@Mothblocks Clothing applying FOV is now componentized, and the FOV handler is now a component too.

**Comment by ATH1909** [comment]
> Once you pick up and wear a gas mask you'll notice a cone behind you.

Does this affect gas mask subtypes, such as sec masks, welding gas masks, clown masks, etc.?

**Comment by Azarak** [comment]
> > Once you pick up and wear a gas mask you'll notice a cone behind you.
> 
> Does this affect gas mask subtypes, such as sec masks, welding gas masks, clown masks, etc.?

All types of gasmask, yes. This will not apply if a visor is lifted up (only some types can do that)
I could go after all gas masks and see which ones got tint and which ones dont and apply it like that aswell should a maintainer desire.

**Comment by tralezab** [comment]
guys i know tint and vision blocking and worries about mining is on your mind but i assure you there will be solutions for those. if we have to make mining mobs ignore fov that's fine

File: code/datums/components/clothing_fov_visor.dm [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+/datum/component/clothing_fov_visor/Initialize(fov_angle)
+	. = ..()
+	if(!isclothing(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/obj/item/clothing/clothing_parent = parent
+	src.fov_angle = fov_angle
+	src.visor_up = clothing_parent.up //Initial values could vary, so we need to get it.
+
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/on_equip)
```
Comment (Mothblocks) [inline code/datums/components/clothing_fov_visor.dm:18]: Put these in RegisterForParent

File: code/modules/mob/living/login.dm [hunk 23]
```diff
@@ -26,3 +26,5 @@
 		changeling.regain_powers()
 
 	med_hud_set_status()
+
+	update_fov_client()
```
Comment (Mothblocks) [inline code/modules/mob/living/login.dm:30]: This is still hitting my concerns of increasing surface area. The FOV component should be listening for the login signal.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Looking better

File: code/modules/mob/living/login.dm (hunk 23 above)
Comment (Azarak) [reply code/modules/mob/living/login.dm:30]: It cant listen for Login because it won't exist at that time

File: code/datums/components/clothing_fov_visor.dm (hunk 22 above)
Comment (Azarak) [reply code/datums/components/clothing_fov_visor.dm:19]: Done

File: code/modules/mob/living/login.dm (hunk 23 above)
Comment (Mothblocks) [reply code/modules/mob/living/login.dm:30]: So is this just being added to every mob, regardless of if it is actually being used (whether through items or config)? It feels like it would already exist if it was actually needed.

Comment (Azarak) [reply code/modules/mob/living/login.dm:30]: The component is only added to the mob if the mob has a field of view and a client. `update_fov_client()` is a proc to check the conditions and update/create the component if the mob is eligible.

**Comment by MrStonedOne** [comment]
> you dont get a breath mask as a miner, you only get mining gas masks in both your survival box and in your suit storage units, you are intended to wear them the entire round



I haven't seen anybody credibly address this point so im gonna keep echoing it until that happens.

**Comment by necromanceranne** [comment]
> guys i know tint and vision blocking and worries about mining is on your mind but i assure you there will be solutions for those. if we have to make mining mobs ignore fov that's fine

Why not keep it and...yeah. Like, the mask has armor, you're fully incentivized into using it already.

**Comment by InvalidArgument3** [comment]
I just thought about binding a key to "spin" constantly, though not sure what the dizziness cooldown is for that. It would be like a radar ping. Pretty amusing

**Comment by optimumtact** [comment]
> > you dont get a breath mask as a miner, you only get mining gas masks in both your survival box and in your suit storage units, you are intended to wear them the entire round
> 
> I haven't seen anybody credibly address this point so im gonna keep echoing it until that happens.

As long as it has it's eye protect removed, it's fine to have it not apply an FoV, this shouldn't have any effect on miners on lavaland.

**Comment by Azarak** [comment]
Clown and Mime masks are excluded from getting the FoV now

**Comment by optimumtact** [comment]
how bad could it be right ha ha

**Comment by SethLafuente** [comment]
It would be better if you didnt have a shadow behind you is kind of annoying 

Great blindness improvements I love them but adding fov to masks is really stupid especially sec mask they don't even block your eyes

**Comment by Azarak** [comment]
> It would be better if you didnt have a shadow behind you is kind of annoying
> 
> Great blindness improvements I love them but adding fov to masks is really stupid especially sec mask they don't even block your eyes

You can adjust the density of the shadow right down to 0 in gameplay preferences
